### PR TITLE
Fix the arity limit of method handle

### PR DESCRIPTION
**Pull request by ChengJin01**
The code is to fix the arity limit of
constructor so as to ensure that argument
slots more than the upper limit are captured
in the arity checking for argument slots.

Signed-off-by: CHENGJin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Peer reviewer:   @bjornvar
Senior reviewer:  @DanHeidinga

**Comment by bjornvar** [comment]
I think there are tests in place for the arity limits. Could you add one for constructor MethodHandles?

**Comment by ChengJin01** [comment]
The changes were verified by the existing test case for the arity limits and JCK tests specific to findConstructor. But I agree it never hurts to add one for our own test.

**Comment by ChengJin01** [comment]
Newly added test cases as follows: (already verified)
test_findConstructor_ArityLimit
test_findConstructor_ArityLimit_TooHigh

The rest of changes is replacing Junit assertion with the equivalent in testNG

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 1]
```diff
@@ -565,75 +491,71 @@ public void test_asSpreader_CrossPackage_Virtual() throws Throwable{
...
 	/******************************
 	 * Tests for MethodHandle arity
 	 * ****************************/
 	/**
+	 * Create MethodHandle with highest possible arity and use findConstructor (253)
+	 * @throws Throwable
+	 */
+	@Test(groups = { "level.extended" })
+	public void test_findConstructor_ArityLimit() throws Throwable {
+		MethodHandle mh = null;
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: nitpick - why assign null here to assign it's real value on the next line?

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 2]
```diff
@@ -1024,33 +902,25 @@ public void test_asFixedArity_SamePackage_Static() throws Throwable {
...
-		try {
-			AssertJUnit.assertEquals(5,mhFix.invoke("a","b","c","d","e"));
-		}
-		catch(WrongMethodTypeException e) {
-			wmtThrown = true;
-		}
+		Assert.assertFalse(mhFix.isVarargsCollector());
 		
-		AssertJUnit.assertTrue(wmtThrown);
+		Assert.assertEquals(mhFix.invoke("a","b","c","d","e"), 5);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: If this is expected to throw, it doesn't need the `Assert.assertEquals`.  The `mhFix.invoke` is sufficient.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 3]
```diff
@@ -1090,33 +960,26 @@ public void test_invokeWithArguments_List_SamePackage_Virtual() throws Throwable
...
 		
-		try {
-			AssertJUnit.assertEquals(6,(int)mh.invokeWithArguments(l));
-		}
-		catch(WrongMethodTypeException e) {
-			wmtThrown = true;
-		}
-		
-		AssertJUnit.assertTrue(wmtThrown);
+		Assert.assertEquals((int)mh.invokeWithArguments(l), 6);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: The `assertEquals` is unnecessary here.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 4]
```diff
@@ -1163,23 +1026,20 @@ public void test_invokeWithArguments_List_CrossPackage_Static() throws Throwable
...
 		MethodHandle mh = MethodHandles.lookup().findVirtual(Object.class, "toString", MethodType.methodType(String.class));
-		try {
-			mh.invokeWithArguments("receiver", null, null);
-			Assert.fail("Should have thrown WMT");
-		} catch(WrongMethodTypeException e) {
-			
-		}
+		mh.invokeWithArguments("receiver", null, null);
+		Assert.fail("Should have thrown WMT");
+		Assert.fail("The test case failed to throw an WrongMethodTypeException in the case of null arguments");
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Two fail calls back to back

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 5]
```diff
@@ -1203,30 +1063,22 @@ public void test_invokeWithArguments_VarargsObj_SamePackage_Virtual_VarargsMetho
...
-		
-		try {
-			AssertJUnit.assertEquals(6,(int)mh.invokeWithArguments(1,2,3));
-		}
-		catch(WrongMethodTypeException e) {
-			wmtThrown = true;
-		}
-		
-		AssertJUnit.assertTrue(wmtThrown);
+		Assert.assertEquals((int)mh.invokeWithArguments(1,2,3), 6);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: The assertEquals isn't required here.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 6]
```diff
@@ -1273,35 +1125,27 @@ public void test_invokeWithArguments_VarargsObj_CrossPackage_Static() throws Thr
...
-		
-		try {
-			AssertJUnit.assertEquals(0,(int)mh2.invokeWithArguments(l2));
-		}
-		catch(WrongMethodTypeException e) {
-			wrongMethodTypeExceptionThrown = true;
-		}
-		
-		AssertJUnit.assertTrue(wrongMethodTypeExceptionThrown);
+		Assert.assertEquals((int)mh2.invokeWithArguments(l2), 0);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: The assertEquals is unnecessary here.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 7]
```diff
@@ -1518,116 +1338,76 @@ public void test_isVarargsCollector_Using_asVarargsCollector_asTyped_arg_int_to_
...
-    	
-    	try {
-    		AssertJUnit.assertEquals(5,handle.invoke(5));
-    	}
-    	catch(WrongMethodTypeException e) {
-    		wmtHit = true;
-    	}
-    	
-    	AssertJUnit.assertTrue(wmtHit);
+    	Assert.assertEquals(handle.invoke(5), 5);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: This will need to be split into it's own test.  Otherwise, this will never be executed as the previous one will throw.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 8]
```diff
@@ -1518,116 +1338,76 @@ public void test_isVarargsCollector_Using_asVarargsCollector_asTyped_arg_int_to_
...
-    	boolean nfeHit = false;
-    	boolean wmtHit = false;
     	
-    	try {
-    		AssertJUnit.assertEquals(5,handle.invoke("s"));
-    	}
-    	catch(NumberFormatException e) {
-    		nfeHit = true;
-    	}
+    	Assert.assertEquals(handle.invoke("s"), 5);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: `assertEquals` not required here/

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 9]
```diff
@@ -1518,116 +1338,76 @@ public void test_isVarargsCollector_Using_asVarargsCollector_asTyped_arg_int_to_
...
-		try{
-			AssertJUnit.assertEquals("[a,b]",(String)mh.invokeExact("a","b"));
-		}
-		catch(WrongMethodTypeException e) {
-			wmtHit = true;
-		}
 		
-		AssertJUnit.assertTrue(wmtHit);
+		Assert.assertEquals((String)mh.invokeExact("a","b"), "[a,b]");
+		Assert.fail("The test case failed to throw an WrongMethodTypeException");
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Everything after this is dead code.  Can you split it into another test?

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 1 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Not sure why the original test case was written in this way. Already fixed.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 2 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Agreed and fixed.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 3 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Fixed at multiple places.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 4 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Updated.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 5 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Fixed.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 6 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Fixed.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 8 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Fixed.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 7 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Already split.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java (hunk 9 above)
Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Already split.

File: test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java [hunk 10]
```diff
@@ -565,75 +491,70 @@ public void test_asSpreader_CrossPackage_Virtual() throws Throwable{
...
 	 * Tests for MethodHandle arity
 	 * ****************************/
 	/**
+	 * Create MethodHandle with highest possible arity and use findConstructor (253)
+	 * @throws Throwable
+	 */
+	@Test(groups = { "level.extended" })
+	public void test_findConstructor_ArityLimit() throws Throwable {
+		MethodHandle mh = MethodHandles.lookup().findConstructor(HelperConstructorClass.class, helperMethodType_void_253int);
+		mh = mh.asSpreader(int[].class, 253);
```
Comment (DanHeidinga) [inline test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: Does this need an assert?  Or is not throwing an exception sufficient?

Comment (ChengJin01) [reply test/Jsr292/src/com/ibm/j9/jsr292/MethodHandleTest.java]: This is a normal test case that should pass in any case; otherwise, any unexpected exception will be captured there. 

In addition, we've got test_findConstructor_ArityLimit_TooHigh() to capture the exception in the corner case.

**Comment by DanHeidinga** [comment]
Jenkins test sanity
